### ISSUE TITLE
Maximum number of UEs and parallelism 

### ISSUE THREAD

**Issue by raoufkh** [aligungr/UERANSIM#359]
Hello @aligungr 

Before anything, I would like to congratulate you for (not the project because it is already done) but for the continuity and the progress of the work.

Would it be possible to have answers to the following questions:
- What is the maximum number of UEs that can be simulated with the command `nr-ue -c myconfig.yaml -n <number>`? Is UERANSIM imposing some limit?
- When simulating multiple UEs with the above command, UE-associated messages will be sent in parallel? I'm asking about that because we want measure the impact running multiple UEs in parallel on the Free5GC. We'll do this using [our Helm charts](https://github.com/Orange-OpenSource/towards5gs-helm/tree/main/charts) which allows customizing the command that will be launched at the startup of the UE container.

Another point to discuss, is about the way to specify IP addresses (local, N3 and N2 ) in the gNB configuration. Do you plan or do you think it is possible to specify the name of the network interfaces to use instead of IP addresses. This will allow us to leverage the scaling capabilities of the Kubernetes platform to run for example multiple gNBs. However, the IP addresses used by the gNB are static at the current version. That implies that the same address will be used by several instances of gNB and therefore it will be impossible to identify each one in the network.  I know in Python there is the pyroute2 library that allows to retrive the IP address of a given interface from the network interface name, but no idea about how to do it in C++.

Regards,
raouf

**Comment by aligungr** [comment]
@raoufkh 

Thanks for the kind words @raoufkh 

---

> What is the maximum number of UEs that can be simulated with the command nr-ue -c myconfig.yaml -n <number>? Is UERANSIM imposing some limit?

- Maximum number of UEs for a single nr-ue process is **512**
- Maximum number of nr-ue process for a single machine is **4 billion**
- Maximum number of UEs for a single gNodeB is **2 billion**.

In other words, with `nr-ue -c myconfig.yaml -n <number>` command the limit is 512, but you can run this command as many as you want. So, in practice it's almost infinity, but most of the time the limitation is determined by the hardware power of the machine.

> When simulating multiple UEs with the above command, UE-associated messages will be sent in parallel?

Yes, exactly. And by the way, if you want to increase the concurrency, you can also use multiple gNodeBs in addition to multiple UEs.

> Do you plan or do you think it is possible to specify the name of the network interfaces to use instead of IP addresses

That's a nice suggestion, I'll implement this feature in the next releases.

Thanks

**Comment by raoufkh** [comment]
Thank you @aligungr for this nice explanation. The answer is very clear to me.

Good luck for the future

**Comment by raoufkh** [comment]
Hi @hxx-98 

Can you show the content of `myconfig.yaml`?

**Comment by hxx-98** [comment]
@raoufkh hi, i can't find myconfig.yaml in UERANSIM

**Comment by raoufkh** [comment]
You have to refer a valid and existing configuration file. @aligungr is providing some sample configuration files in [this path](https://github.com/aligungr/UERANSIM/tree/master/config)
